[WIP] #430 make e2e csv full report all backends target

emabota opened the pull request:
This PR is aimed at the issue #430. 
It focus, initially, on the implementation of the e2e_test in go. 
I decided to start by aiming at a program that will work on this conditions: **ip_family=ipv4, backend=iptables, cluster_count=1**

Studying the current `test_e2e.sh` script I identified the main functions required to have a working `test_e2e.go` program. And documented it in [here](https://docs.google.com/document/d/1fanr95mp9Sc-wh9Ss0PFqu9yqUA6tdA2fRxUWq3lmmY/edit?usp=sharing). Please check the MILESTONES on the document, and let me know if there is something missing. 

So basically my current approach is to implement each identified function, in **go**. 
On this PR I have implemented the first 5 functions: `Info_error_exit, set_sysctl, set_host_networking_settings, add_to_path, setup_kind`

@jayunit100 , @rajaskakodkar please comments. Let me know if this is the way to go.

emabota committed:
Initial commit. Implementation of the sysctl and if_error_exit function

emabota committed:
Implementation of the set_host_network_settings function

emabota committed:
Implementing the add_to_path function - WIP.

emabota committed:
Implemented the add_to_path function.

emabota committed:
Implementation of the add_to_path function - changed the way to check if the directory path exist in the /home/emabota/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/go/bin:/home/emabota/go/bin env variable.

emabota committed:
setup_kind func: implemented the it's functionality. The live display of Curl, on the terminal, of the ongoind download is missing

jayunit100 commented:
Thanks can you attach the cover sample output to a README.report.md file or something like that in this dir?

emabota committed:
setup_kind: Fixed the if statement. It only istalls kind if there is no kind in the system

emabota committed:
Added the folder temp_go

emabota committed:
Implemented the setup_kubectl function.

emabota committed:
Implementation of the setup_ginkgo function

emabota committed:
Implementation of the function setup_bpf2go.

emabota committed:
Implementation of the functions verify_host_network_settings and verify_sysctl_setting

emabota commented:
I don't understand why this is failing - https://github.com/kubernetes-sigs/kpng/actions/runs/4480600683/jobs/7876158548

emabota committed:
Implementation of the detect_container_engine function

jayunit100 commented:
https://github.com/kubernetes/kubernetes/issues/108891 <-- the failure is a flake.  just push a readme commit or something to rerun the testss

emabota committed:
This is a meaningless commit, just to trigger the CI tests again in order to pass some flakes!

emabota committed:
container_build function - ongoing implementation. Can't build docker images - debugging

emabota committed:
container_build function - fixed the the issue that was creating the failure of the docker build. And finished the implementation

emabota committed:
Ongoing implementation of creation_cluster - done with the creation of the kind-config.yaml file. Implementation of set_e2e_dir function

emabota committed:
Conclusion of the create_cluster function

emabota committed:
Implementation of the install_kpng function.

jayunit100 commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+	fmt.Printf("Cluster %s is created.\n", cluster_name)
+
+}
+
```
godocccc

emabota replied:
Done

jayunit100 commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+	}
+	fmt.Printf("%s is operational.\n",cluster_name)
+}
+
```
add godoc

```
// install_kpng...
```

emabota replied:
Done

knabben replied:
please mark those godoc thread as resolved

jayunit100 commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+	fmt.Println("Installation of kpng is done.")
+
+}
+
```
godoc

emabota replied:
Done

jayunit100 commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+package main
+
```
https://go.dev/blog/godoc

emabota replied:
Done

daman1807 commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+	fmt.Println("Detected Container Engine:", CONTAINER_ENGINE)
+}
+
+func container_build(CONTAINER_FILE string, ci_mode bool) {
```
best to use `containerFile` and `ciMode` as parameters, following go conventions

https://go.dev/doc/effective_go#:~:text=Finally%2C%20the%20convention%20in%20Go%20is%20to%20use%20MixedCaps%20or%20mixedCaps%20rather%20than%20underscores%20to%20write%20multiword%20names.

emabota replied:
Done

daman1807 commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+	if err != nil {
+		log.Fatalf("Cannot patch kube-proxy.\n", err)
+	}
+	fmt.Println("Kube-proxy patched! Guys how can I test this? To find out if it was really successful.")
```
"Guys how can I test this? To find out if it was really successful." 😅 
I guess if the `err` is nil, the cmd executed successfully.

daman1807 commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+
+	// Setting vars for generate the kpng deployment based on template
+	_ = os.Setenv("kpng_image", KPNG_IMAGE_TAG_NAME) 
+    _ = os.Setenv("image_pull_policy", "IfNotPresent") 
```
identation

emabota replied:
@aroradaman I don't know why this is so. On my repo there is no identation issues  https://github.com/emabota/kpng/blob/8f0f3eb5229fee8eca3ff0650c5946d91d75582b/hack/test_e2e.go#L708
Any idea what's causing this?

knabben replied:
use `go fmt hack/test_e2e.go`

emabota committed:
Implementation of the function run_tests. Added comments to all the functions based on Go Doc Comments conventions.

emabota committed:
Changed the Constants names following the Go Conventions.

emabota committed:
Changed the parameters names following the Go Conventions.

emabota committed:
Changed the functions names following the Go Conventions.

emabota committed:
Changed the variables names in all functions following Go Conventions.

emabota committed:
Changed the variable kpngDir to kpngPath. Added comments to the package main and the global variables kpngPath and containerEngine.

emabota committed:
Added comments to the function commandExist about the Bash builtin command - command.

emabota committed:
Updated and added some comments. Changed the names of some variables and paramenters.

addToPath() - Wrote a clear function comment.
setSysctl() - changed the variable variableValue to attributeValue.
verifySysctlSetting() - added ifErrorExit(err) for the strconv.
setupKind() - changed the TODO comment just to make it more clear.
setupGinkgo() - just fixed some commas in the comment.
installBinaries() - changed the parameter name from baseDirPath to hackDirectory. Added comments to the function. Changed the variable name baseDir to hackDir.

emabota committed:
Updated some comments. And added some TODOs

containerBuild() removed an unnecessary comment
containerBuild() Added a TODO task. Added some clarity to the function comment
setE2eDir() added a TODO task.
setE2eDir() added some comments about the function
createCluster() added comments about the cmdStringSet variable
installKpng() fixed some typos

jayunit100 commented:
/approve

emabota commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+	if err != nil && os.IsNotExist(err) {
+		if_error_exit(err)
+	}
+
```
Hi @aroradaman, after reading the link - [](https://pkg.go.dev/k8s.io/klog/v2) - I still don't understand why we should use logger instead of Println. At first I thought you were talking about the **Log** package! But then I realized that it's an all different package! Please can you help me understanding why **klog** is a better option? And wouldn't **Log** work here?

aroradaman replied:
@emabota klog will give you the option for structured logging and you can control the verbosity ... also the logging will be consistent through kpng

emabota commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,944 @@
+
+}
+
+func create_cluster(cluster_name, bin_dir, ip_family, artifacts_directory string, ci_mode bool) {
```
When kube-proxy is disabled the coredns pods FAIL to start! 
`$ kubectl get pods -n kube-system 
NAME                                                           READY   STATUS    RESTARTS   AGE
coredns-565d847f94-6chxd                                       0/1     Running   0          16m
coredns-565d847f94-86df5                                       0/1     Running   0          16m
etcd-kpng-e2e-ipv4-iptables-control-plane                      1/1     Running   0          16m
kindnet-f7fmc                                                  1/1     Running   0          15m
kindnet-fwtqn                                                  1/1     Running   0          16m
kindnet-n56n2                                                  1/1     Running   0          15m
kube-apiserver-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
kube-controller-manager-kpng-e2e-ipv4-iptables-control-plane   1/1     Running   0          16m
kube-scheduler-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
`

`$ kubectl get endpoints --all-namespaces 
NAMESPACE     NAME         ENDPOINTS         AGE
default       kubernetes   172.18.0.2:6443   16m
kube-system   kube-dns                       15m
`

emabota replied:
```
$ kubectl get pods -n kube-system 
NAME                                                           READY   STATUS    RESTARTS   AGE
coredns-565d847f94-6chxd                                       0/1     Running   0          16m
coredns-565d847f94-86df5                                       0/1     Running   0          16m
etcd-kpng-e2e-ipv4-iptables-control-plane                      1/1     Running   0          16m
kindnet-f7fmc                                                  1/1     Running   0          15m
kindnet-fwtqn                                                  1/1     Running   0          16m
kindnet-n56n2                                                  1/1     Running   0          15m
kube-apiserver-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
kube-controller-manager-kpng-e2e-ipv4-iptables-control-plane   1/1     Running   0          16m
kube-scheduler-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
```

emabota replied:
```
$ kubectl get endpoints --all-namespaces 
NAMESPACE     NAME         ENDPOINTS         AGE
default       kubernetes   172.18.0.2:6443   16m
kube-system   kube-dns                       15m

```

knabben commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,917 @@
+/*
+Test_e2e performs tests on the Kubernetes cluster. 
+It creates a kubernetes cluster based in kind + kpng, and performs ginkgo tests on it. 
+
+Usage:
```
header license missing

knabben commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,917 @@
+	cmd := exec.Command("bash", "-c", cmdScript)
+	err := cmd.Run()
+	
+	if err == nil {
```
`return err == nil`

knabben commented on hack/test_e2e.go:
```diff
@@ -0,0 +1,917 @@
+		log.Fatal(err)
+	}
+	
+	if ciMode == true {
```
`if ciMode {`

emabota commented:
/remove-lifecycle stale

rajaskakodkar commented:
@emabota just checking if you are still working on this

emabota commented:
> @emabota just checking if you are still working on this

Hi @rajaskakodkar , yes, I am still interested. Just had a tight schedule lately, but have planned to resume it on June.
